### PR TITLE
added hot reload support for stream aggregation configs (#3969)

### PR DESCRIPTION
**Pull request by Amper**
issue #3969

### CODE REVIEW

File: app/vmagent/remotewrite/remotewrite.go [hunk 1]
```diff
@@ -585,6 +623,38 @@ func (rwctx *remoteWriteCtx) pushInternal(tss []prompbmarshal.TimeSeries) {
...
+
+	if rwctx.cfgHash != cfgHash {
+		err = rwctx.sas.ReInitConfigs(cfgs)
+		if err != nil {
+			logger.Errorf("cannot apply new stream aggregation configs from file %q: %s", rwctx.sasFile, err)
+			rwctx.saCfgSuccess.Set(0)
+			rwctx.saCfgReloadErr.Inc()
+			return err
+		}
+	}
```
Comment (hagen1778) [inline app/vmagent/remotewrite/remotewrite.go:649]:
```suggestion
	if rwctx.cfgHash == cfgHash {
	  return nil
	}
	err = rwctx.sas.ReInitConfigs(cfgs)
	if err != nil {
		logger.Errorf("cannot apply new stream aggregation configs from file %q: %s", rwctx.sasFile, err)
		rwctx.saCfgSuccess.Set(0)
		rwctx.saCfgReloadErr.Inc()
		return err
	}
```

File: app/vmagent/remotewrite/remotewrite.go [hunk 2]
```diff
@@ -451,13 +469,22 @@ type remoteWriteCtx struct {
 	c   *client
 
 	sas                 *streamaggr.Aggregators
+	sasFile             string
 	streamAggrKeepInput bool
+	dedupInterval       time.Duration
```
Comment (hagen1778) [inline app/vmagent/remotewrite/remotewrite.go:474]: how this field is used?

File: app/vmagent/remotewrite/remotewrite.go [hunk 3]
```diff
@@ -168,29 +168,47 @@ func Init() {
...
+					}
+				}
+			}
+			for _, rwctx := range rwctxsDefault {
+				err = rwctx.reloadStreamAggr()
+				if err != nil {
+					hasErrors = true
+				}
+			}
+			if !hasErrors {
```
Comment (hagen1778) [inline app/vmagent/remotewrite/remotewrite.go:200]: I'm not sure if we should allow re-loading configs for one URL when config reload for another URL failed.
While it looks cool, but it is not consistent with relableingRules reload and moves responsibility for logging and metrics from the caller to the function it calls. This is also inconsistent with relabeling.
For example, `reloadStreamAggr` returns an `error`. But no-one ever reads that error. That's a sign of inconsistency.

File: lib/streamaggr/streamaggr.go [hunk 4]
```diff
@@ -172,11 +208,71 @@ func (a *Aggregators) Push(tss []prompbmarshal.TimeSeries) {
...
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+
 	for _, aggr := range a.as {
 		aggr.Push(tss)
 	}
 }
 
+// ReInitConfigs reinits state of Aggregators a with the given new stream aggregation config
+func (a *Aggregators) ReInitConfigs(cfgs []*Config) error {
```
Comment (hagen1778) [inline lib/streamaggr/streamaggr.go:221]: would you mind covering this func with tests?

**Review by hagen1778 — CHANGES_REQUESTED** [review]

File: app/vmagent/remotewrite/remotewrite.go [hunk 5]
```diff
@@ -168,29 +168,47 @@ func Init() {
...
-			configSuccess.Set(1)
-			configTimestamp.Set(fasttime.UnixTimestamp())
+			relabelConfigSuccess.Set(1)
+			relabelConfigTimestamp.Set(fasttime.UnixTimestamp())
 			logger.Infof("Successfully reloaded relabel configs")
+
+			hasErrors := false
+			for _, rwctxs := range rwctxsMap {
+				for _, rwctx := range rwctxs {
+					err = rwctx.reloadStreamAggr()
```
[the remark this answers is not in the document]
Comment (hagen1778) [reply app/vmagent/remotewrite/remotewrite.go:188]: you can verify it by adding test with concurrent config reload. Something like here https://github.com/VictoriaMetrics/VictoriaMetrics/blob/8fdd613f257eaa1f6ccef5236f6f855441a3fc54/app/vmalert/manager_test.go#LL39

**Review by f41gh7 — COMMENTED** [review]
Overall looks good. See comments.

 Also, it would be good to have some tests.

File: app/vmagent/remotewrite/remotewrite.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Amper) [reply app/vmagent/remotewrite/remotewrite.go]: I made without using rwctxsMap (and without locks)

File: app/vmagent/remotewrite/remotewrite.go (hunk 2 above)
Comment (Amper) [reply app/vmagent/remotewrite/remotewrite.go]: fixed

File: app/vmagent/remotewrite/remotewrite.go (hunk 3 above)
Comment (Amper) [reply app/vmagent/remotewrite/remotewrite.go]: fixed

File: lib/streamaggr/streamaggr.go (hunk 4 above)
Comment (Amper) [reply lib/streamaggr/streamaggr.go:219]: done

File: app/vmagent/remotewrite/remotewrite.go [hunk 6]
```diff
@@ -168,29 +168,47 @@ func Init() {
...
-
 			allRelabelConfigs.Store(rcs)
-			configSuccess.Set(1)
-			configTimestamp.Set(fasttime.UnixTimestamp())
+			relabelConfigSuccess.Set(1)
+			relabelConfigTimestamp.Set(fasttime.UnixTimestamp())
 			logger.Infof("Successfully reloaded relabel configs")
+
+			hasErrors := false
+			for _, rwctxs := range rwctxsMap {
```
[the remark this answers is not in the document]
Comment (Amper) [reply app/vmagent/remotewrite/remotewrite.go]: fixed

File: app/vmagent/remotewrite/remotewrite.go [hunk 7]
```diff
@@ -168,29 +168,47 @@ func Init() {
...
+				for _, rwctx := range rwctxs {
+					err = rwctx.reloadStreamAggr()
+					if err != nil {
+						hasErrors = true
```
[the remark this answers is not in the document]
Comment (Amper) [reply app/vmagent/remotewrite/remotewrite.go]: fixed

File: lib/streamaggr/streamaggr.go [hunk 8]
```diff
@@ -127,9 +143,23 @@ type Config struct {
...
+	if err != nil {
+		return 0, fmt.Errorf("cannot marshal stream aggregation rule %+v: %w", cfg, err)
+	}
+	return xxhash.Sum64(data), nil
+}
+
 // Aggregators aggregates metrics passed to Push and calls pushFunc for aggregate data.
 type Aggregators struct {
-	as []*aggregator
+	mu            sync.RWMutex
```
[the remark this answers is not in the document]
Comment (Amper) [reply lib/streamaggr/streamaggr.go]: done

File: app/vmagent/remotewrite/streamagg.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		}
+	}
+
+	saCfgSuccess.Set(1)
+	saCfgTimestamp.Set(fasttime.UnixTimestamp())
+	return nil
+}
+
+// GetCurrentConfig returns the current stream aggregation config with the given idx.
+func (r *SaConfigLoader) GetCurrentConfig(idx int) SaConfigRules {
```
Comment (f41gh7) [inline app/vmagent/remotewrite/streamagg.go:103]: This code may panic.

 I suggest to add the following check:

if len(all) == 0 {
    return nil
}
if argIdx >= len(all) {
		if len(all) == 1 {
			return all[0]
		}
		return ""
	}

File: app/vmagent/remotewrite/remotewrite.go [hunk 10]
```diff
@@ -510,15 +515,31 @@ func newRemoteWriteCtx(argIdx int, at *auth.Token, remoteWriteURL *url.URL, maxI
...
 		rwctx.streamAggrKeepInput = streamAggrKeepInput.GetOptionalArg(argIdx)
+
+		configReloaderWG.Add(1)
+		go func() {
+			defer configReloaderWG.Done()
+			for {
+				select {
+				case <-stopCh:
+					return
+				case c := <-saConfigLoader.UpdatesCh(argIdx): // wait for updates after config reload
```
Comment (f41gh7) [inline app/vmagent/remotewrite/remotewrite.go:536]: I think, dymanic updates is not needed.

 `remoteWrite.streamAggr.config` - always has the same number of files and it cannot be changed dynamically.

**Review by f41gh7 — APPROVED** [review]
LGTM

**Pull request merged**